### PR TITLE
Add exact event checking when converting event-values in last attempt

### PR DESCRIPTION
**Pull request by TheLimeGlass**
### Description
When Skript can't find an event value that matches the requesting event OR type. It starts attempting to look at other event values and checks if the type is assignable then checks if the event is assignable, but Skript will error in the rare case that the event value is actually of a GREATER subclass than the requesting event of the event-value. In the case of #5016 Skript was grabbing the newly added CraftItemEvent event value (CraftItemEvent extends InventoryClickEvent) rather than InventoryClickEvent event value for ItemStack when requesting an ItemType. So Skript wanted CraftItemEvent ItemStack event value to convert it ItemType but the event was InventoryClickEvent so it threw a method not found error.

There is really only a new loop added, GitHub makes the changes look more than it is.

Since this is purely runtime, a test can't be captured for this until my JUnit test pull request (https://github.com/SkriptLang/Skript/pull/4979) I have made a comment in that pull request to add a regression test for this exact issue.

This will be cherry picked to 2.6.4.

---
**Target Minecraft Versions:** any
**Related Issues:** https://github.com/SkriptLang/Skript/issues/5016 and https://github.com/SkriptLang/Skript/issues/5032

### CODE REVIEW

**Review by APickledWalrus — APPROVED** [review]

File: src/main/java/ch/njol/skript/registrations/EventValues.java [hunk 1]
```diff
@@ -271,58 +271,79 @@ public T get(E event) {
 				}
 			};
 		}
-		// Fourth check will attempt to convert the event value to the type.
+		// Fourth check will attempt to convert the event value to the requesting type.
+		// This first for loop will check that the events are exact. See issue #5016
 		for (EventValueInfo<?, ?> ev : eventValues) {
-			boolean checkInstanceOf = !ev.event.isAssignableFrom(e);
-			if (checkInstanceOf && !e.isAssignableFrom(ev.event))
+			if (!event.isInstance(ev.event))
```
Comment (TPGamesNL) [inline src/main/java/ch/njol/skript/registrations/EventValues.java:277]: `event.isInstance(ev.event)` is always false: `event is a Class<? extends Event>` so isInstance can only be true if called with an event instance. `ev.event` is a class.

**Review by TPGamesNL — CHANGES_REQUESTED** [review]

File: src/main/java/ch/njol/skript/registrations/EventValues.java [hunk 2]
```diff
@@ -271,58 +271,79 @@ public T get(E event) {
 				}
 			};
 		}
-		// Fourth check will attempt to convert the event value to the type.
+		// Fourth check will attempt to convert the event value to the requesting type.
+		// This first for loop will check that the events are exact. See issue #5016
 		for (EventValueInfo<?, ?> ev : eventValues) {
-			boolean checkInstanceOf = !ev.event.isAssignableFrom(e);
-			if (checkInstanceOf && !e.isAssignableFrom(ev.event))
+			if (!event.getClass().equals(ev.event))
```
Comment (TPGamesNL) [inline src/main/java/ch/njol/skript/registrations/EventValues.java:277]: `event` is already a `Class<? extends Event>`, calling getClass gives a `Class<Class>` so the equals never returns true.

**Review by TPGamesNL — CHANGES_REQUESTED** [review]

File: src/main/java/ch/njol/skript/registrations/EventValues.java [hunk 3]
```diff
@@ -253,10 +253,10 @@ public T get(E event) {
 		for (EventValueInfo<?, ?> ev : eventValues) {
```
Comment (AyhamAl-Ali) [inline src/main/java/ch/njol/skript/registrations/EventValues.java:253]: `ev` and `event` are very similar, can we use `evi` or `info` like in other places?
```suggestion
		for (EventValueInfo<?, ?> evi : eventValues) {
```

File: src/main/java/ch/njol/skript/registrations/EventValues.java [hunk 4]
```diff
@@ -211,118 +211,139 @@ public static <T, E extends Event> T getEventValue(E e, Class<T> c, int time) {
 	 * @see EventValueExpression#EventValueExpression(Class)
 	 */
 	@Nullable
-	public static <T, E extends Event> Getter<? extends T, ? super E> getEventValueGetter(Class<E> e, Class<T> c, int time) {
-		return getEventValueGetter(e, c, time, true);
+	public static <T, E extends Event> Getter<? extends T, ? super E> getEventValueGetter(Class<E> event, Class<T> c, int time) {
```
Comment (TPGamesNL) [inline src/main/java/ch/njol/skript/registrations/EventValues.java:214]: Rename javadoc `@param e` as well

File: src/main/java/ch/njol/skript/registrations/EventValues.java [hunk 5]
```diff
@@ -211,118 +211,139 @@ public static <T, E extends Event> T getEventValue(E e, Class<T> c, int time) {
...
+			Getter<? extends T, ? super E> getter = (Getter<? extends T, ? super E>) getConvertedGetter(eventValueInfo, c, true);
 			if (getter == null)
 				continue;
 			
-			if (!checkExcludes(ev, e))
+			if (!checkExcludes(eventValueInfo, event))
+				return null;
+			return getter;
+		}
+		// This second loop for the fourth check will attempt to look for converters assignable to the class of the provided event.
```
Comment (TPGamesNL) [inline src/main/java/ch/njol/skript/registrations/EventValues.java:288]: This is now the fifth check, not fourth.

File: src/main/java/ch/njol/skript/registrations/EventValues.java [hunk 6]
```diff
@@ -211,118 +211,139 @@ public static <T, E extends Event> T getEventValue(E e, Class<T> c, int time) {
...
-			boolean checkInstanceOf = !ev.event.isAssignableFrom(e);
-			if (checkInstanceOf && !e.isAssignableFrom(ev.event))
+		// Fourth check will attempt to convert the event value to the requesting type.
+		// This first for loop will check that the events are exact. See issue #5016
+		for (EventValueInfo<?, ?> eventValueInfo : eventValues) {
+			if (!event.equals(eventValueInfo.event))
 				continue;
 			
-			Getter<? extends T, ? super E> getter = (Getter<? extends T, ? super E>) getConvertedGetter(ev, c, checkInstanceOf);
+			Getter<? extends T, ? super E> getter = (Getter<? extends T, ? super E>) getConvertedGetter(eventValueInfo, c, true);
```
Comment (TPGamesNL) [inline src/main/java/ch/njol/skript/registrations/EventValues.java:280]:
```suggestion
			Getter<? extends T, ? super E> getter = (Getter<? extends T, ? super E>) getConvertedGetter(eventValueInfo, c, false);
```
No need to check instanceof if the events equal

File: src/main/java/ch/njol/skript/registrations/EventValues.java [hunk 7]
```diff
@@ -211,118 +211,139 @@ public static <T, E extends Event> T getEventValue(E e, Class<T> c, int time) {
...
+				return null;
+			return getter;
+		}
+		// This second loop for the fourth check will attempt to look for converters assignable to the class of the provided event.
+		for (EventValueInfo<?, ?> eventValueInfo : eventValues) {
+			// The requesting event must be assignable to the event value's event. Otherwise it'll throw an error.
+			if (!event.isAssignableFrom(eventValueInfo.event))
+				continue;
+			
+			Getter<? extends T, ? super E> getter = (Getter<? extends T, ? super E>) getConvertedGetter(eventValueInfo, c, false);
```
Comment (TPGamesNL) [inline src/main/java/ch/njol/skript/registrations/EventValues.java:294]:
```suggestion
			Getter<? extends T, ? super E> getter = (Getter<? extends T, ? super E>) getConvertedGetter(eventValueInfo, c, true);
```
(no need for the special condition anymore with the previous loop)

File: src/main/java/ch/njol/skript/registrations/EventValues.java [hunk 8]
```diff
@@ -211,118 +211,139 @@ public static <T, E extends Event> T getEventValue(E e, Class<T> c, int time) {
...
 	}
-	
+
+	/**
+	 * Return a converter wrapped in a getter that will grab the requested value by converting from the given event value info.
+	 * 
+	 * @param info The event value info that will be used to grab the value from
+	 * @param to The class that the converter will look for to convert the type from the event value to
+	 * @param checkInstanceOf If the event must be an exact instance of the event value info's event or not.
+	 * @return The found Converter wrapped in a Getter object.
```
Comment (TPGamesNL) [inline src/main/java/ch/njol/skript/registrations/EventValues.java:333]:
```suggestion
	 * @return The found Converter wrapped in a Getter object, or null if no Converter was found.
```

**Review by TPGamesNL — CHANGES_REQUESTED** [review]

File: src/main/java/ch/njol/skript/registrations/EventValues.java (hunk 5 above)
Comment (TheLimeGlass) [reply src/main/java/ch/njol/skript/registrations/EventValues.java:288]: It's doing the same thing as the fourth check, a new check would mean doing something different than convertering.

Comment (TPGamesNL) [reply src/main/java/ch/njol/skript/registrations/EventValues.java:288]: It's not the same check: first for exact event match, then for assignable event check.

If you used that approach for all checks, we'd have 2 (maybe 3) in total. It's clearer to make them distinct.

File: src/main/java/ch/njol/skript/registrations/EventValues.java (hunk 6 above)
Comment (TheLimeGlass) [reply src/main/java/ch/njol/skript/registrations/EventValues.java:280]: The check on line 294 would need to be true, adjusted accordingly.

**Review by TPGamesNL — APPROVED** [review]

**Pull request merged**